### DISCUSSION TITLE
Data Importer authorize responds with "empty"

### DISCUSSION THREAD

**Discussion by adiba** [firefly-iii/firefly-iii#9464] [Firefly III]
### Support guidelines

- [x] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [x] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [x] ... there are no open or closed issues that are related to my problem
- [x] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

Using docker.  
I just installed firefly + the data importer, modified the config, set the client id with callback url `http://localhost:81/callback`.

Visiting `http://localhost:81/` shows an authorize button. Clicking it just shows the word "empty" on the page, the url being `http://localhost:81/callback?code=...`

Going to `http://localhost:81/` shows the authorize button again.

### Debug information

Debug information for Firefly III Data Importer version **1.5.7**.

<table><tr><th colspan="2">System information</th></tr><tr><th>Item</th><th>Value</th></tr><tr><td>Version</td><td>1.5.7</td></tr><tr><td>Build</td><td><span>#</span>677, base <span>#</span>92</td></tr><tr><td>System</td><td>PHP 8.3.13, Linux, apache2handler</td></tr></table><table><tr><th colspan="2">App information</th></tr><tr><th>Item</th><th>Value</th></tr><tr><td>Environment</td><td>local</td></tr><tr><td>Debug mode</td><td>false, cache 'file'</td></tr><tr><td>Log level</td><td>debug, stack</td></tr><tr><td>Display errors</td><td>Off, ALL errors</td></tr><tr><td>BCscale</td><td>12</td></tr><tr><td>Trusted proxies</td><td></td></tr></table>

### Expected behaviour

Authorization successful

**Comment by JC5** [comment]
Hi, thanks for opening an issue. 

The redirect is supposed to go to Firefly III, so (presumably) port 80. Not 81. Please check your configuration.

**Comment by adiba** [comment]
> Hi, thanks for opening an issue.
> 
> The redirect is supposed to go to Firefly III, so (presumably) port 80. Not 81. Please check your configuration.

Thank you for your quick response!

Which redirect?  
Visiting :81 redirects me to the authorization page, which is on :80

**Comment by jflorido94** [comment]
Im having the same problem, and I have the container of firefly at 8080 port

**Comment by JC5** [comment]
Could you share your docker configuration please?

**Reply by adiba** [reply]
It is this exact file, downloaded, not copied. I double checked by comparing the contents:

https://raw.githubusercontent.com/firefly-iii/docker/main/docker-compose-importer.yml

**Reply by JC5** [reply]
Cool, could you share the following configuration variables from `.importer.env`?

- FIREFLY_III_URL
- VANITY_URL

**Reply by adiba** [reply]
[.importer.env.txt](https://github.com/user-attachments/files/17736714/default.importer.env.txt)
[.env.txt](https://github.com/user-attachments/files/17736715/default.env.txt)
[.db.env.txt](https://github.com/user-attachments/files/17736716/default.db.env.txt)

`VANITY_URL=http://localhost`
`FIREFLY_III_URL=http://app:8080`

**Reply by JC5** [reply]
Alright, I've built the exact same setup and it works for me ™️ .

- The data importer running on localhost:81 reports its callback url is `http://localhost:81/callback`
- After registering at localhost (Firefly III), I create one asset account.
- I create a client on the Firefly III side, ID=2, and i deselect the "confidential" checkbox.
- I enter "2" in the data importer. I get redirected to localhost for authorization and it sends me back.

Only tip I have is to browse to http://localhost:81/flush to clear out any cookies.

**Reply by adiba** [reply]
How do I change the client ID for the importer? Removing it from the `.importer.env ` does not do it for me. At least I am not asked for a client ID anymore.

**Reply by adiba** [reply]
Where does the "empty\n" response come from?

**Reply by JC5** [reply]
If you change settings, browse to /flush. 

As for the "empty" response, I don't know. I don't believe it's my code (although it has to be lol)

**Reply by cpauls** [reply]
Hi, 
I got the response "empty" too. This error is raised on fresh installations, where no asset accounts are available. If you create the asset account first, the error is not raised. 

regards

**Comment by Armgd** [comment]
Hello,

I'm running into a similar issue. I'm not running in full local but on a device on my home network (FireflyIII version 6.1.22).
Setup is Docker and Traefik as reverse proxy:

### docker compose
```yaml
services:
  app:
    image: fireflyiii/core:latest
    container_name: firefly_iii_core
    restart: always
    labels:
      - traefik.enable=true
      - traefik.http.routers.firefly-app.rule=Host(`firefly.raspi5.home.arpa`)
      - traefik.http.routers.firefly-app.entrypoints=web
      - traefik.http.services.firefly-app.loadbalancer.server.port=8080
      - homepage.group=finance
      - homepage.name=FireflyIII
    volumes:
      - firefly_iii_upload:/var/www/html/storage/upload
    env_file: .env
    networks:
      - firefly_iii
      - traefik_proxy
    depends_on:
      - db
  importer:
    image: fireflyiii/data-importer:latest  
    hostname: importer
    restart: always
    container_name: firefly_iii_importer
    networks:
      - firefly_iii
      - traefik_proxy
    labels:
      - traefik.enable=true
      - traefik.http.routers.firefly-importer.rule=Host(`firefly-import.raspi5.home.arpa`) 
      - traefik.http.routers.firefly-importer.entrypoints=web
      - traefik.http.services.firefly-importer.loadbalancer.server.port=8080
    depends_on:
      - app
    env_file: .importer.env
  redis:
    image: redis:7.4.1-bookworm
    labels:
      - "traefik.tcp.services.redis.loadbalancer.server.port=6379"
    container_name: firefly_iii_redis
    networks:
      - firefly_iii
  db:
    image: mariadb:lts
    hostname: db
    container_name: firefly_iii_db
    restart: always
    env_file: .db.env
    networks:
      - firefly_iii
    volumes:
      - firefly_iii_db:/var/lib/mysql
  cron:
    #
    # To make this work, set STATIC_CRON_TOKEN in your .env file or as an environment variable and replace REPLACEME below
    # The STATIC_CRON_TOKEN must be *exactly* 32 characters long
    #
    image: alpine
    restart: always
    container_name: firefly_iii_cron
    command: sh -c "echo \"0 3 * * * wget -qO- http://app:8080/api/v1/cron/<something>;echo\" | crontab - && crond -f -L /dev/stdout"
    networks:
      - firefly_iii

volumes:
   firefly_iii_upload:
   firefly_iii_db:

networks:
  firefly_iii:
    driver: bridge
  traefik_proxy:
    external: true

```
### Checking everything is running and ports opened

```
26d985172cd2   fireflyiii/core:latest                    "/usr/local/bin/entr…"    7 minutes ago    Up 7 minutes (healthy)   8080/tcp                                                                                                              firefly_iii_core
c53c9ea7331c   fireflyiii/data-importer:latest           "/usr/local/bin/entr…"    13 minutes ago   Up 7 minutes (healthy)   8080/tcp                                                                                                              firefly_iii_importer
3dcc504a28eb   mariadb:lts                               "docker-entrypoint.s…"    2 hours ago      Up 2 hours               3306/tcp                                                                                                              firefly_iii_db
7e80b921e685   redis:7.4.1-bookworm                      "docker-entrypoint.s…"    2 hours ago      Up 2 hours               6379/tcp                                                                                                              firefly_iii_redis
65e2c9ab887a   alpine                                    "sh -c 'echo \"0 3 * …"   2 hours ago      Up 2 hours                                                               
```

### Steps reproduced
Then I go to http://firefly.raspi5.home.arpa, login -> Profile -> Oauth, create new client with ID XX.
In .importer.env, I set 

```
FIREFLY_III_URL=http://app:8080
VANITY_URL=http://firefly.raspi5.home.arpa
FIREFLY_III_CLIENT_ID=XX
TRUSTED_PROXIES=**
```

In .env I have also:
```
TRUSTED_PROXIES=**
```

Then when I go to firefly-import.raspi5.home.arpa:
![image](https://github.com/user-attachments/assets/68a67516-4e1a-410b-b1c6-a121277153f1)

Then when clicking on Authorise:
![image](https://github.com/user-attachments/assets/457db602-3c47-4dfd-a4cb-5f06cb6da9f0)
I got the same error.


### What the logs are saying?
Curious, I wnet into the log of the importer container and something is bothering me:
```
[2024-11-23 15:19:48] local.DEBUG: Now at App\Http\Controllers\TokenController::index  
[2024-11-23 15:19:48] local.DEBUG: No access token in hasAccessToken() session, will return config variable.  
[2024-11-23 15:19:48] local.DEBUG: No client id in hasClientId() session, will return config variable.  
[2024-11-23 15:19:48] local.DEBUG: No base url in getBaseUrl() session, will return config variable.  
[2024-11-23 15:19:48] local.DEBUG: No vanity url in getVanityUrl() session, will return config variable.  
[2024-11-23 15:19:48] local.INFO: The following configuration information was found:  
[2024-11-23 15:19:48] local.INFO: Personal Access Token: "" (limited to 25 chars if present)  
[2024-11-23 15:19:48] local.INFO: Client ID            : "17"  
[2024-11-23 15:19:48] local.INFO: Base URL             : "http://app:8080"  
[2024-11-23 15:19:48] local.INFO: Vanity URL           : "http://firefly.raspi5.home.arpa"  
[2024-11-23 15:19:48] local.DEBUG: Found client ID "17" + URL "http://app:8080" in config, redirect to Firefly III for permission.  
[2024-11-23 15:19:48] local.DEBUG: Now in App\Http\Controllers\TokenController::redirectForPermission(request, "http://app:8080", "http://firefly.raspi5.home.arpa", 17)  
[2024-11-23 15:19:48] local.DEBUG: Query parameters are {"client_id":17,"redirect_uri":"http://firefly-import.raspi5.home.arpa/callback","response_type":"code","scope":"","state":"rs6o7IScd722pNQoJFd6Lu4szVEpcICbIyipyMm1","code_challenge":"YuWGN_e96BnyqBp6WB_WKpQv6QuW5oDF2PPTBM5ONnE","code_challenge_method":"S256"} 
[2024-11-23 15:19:48] local.DEBUG: Now redirecting to "http://firefly.raspi5.home.arpa/oauth/authorize?" (params omitted)  
172.19.0.15 - - [23/Nov/2024:15:19:48 +0100] "GET /token HTTP/1.1" 302 2879 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
127.0.0.1 - - [23/Nov/2024:15:19:52 +0100] "GET /health HTTP/1.1" 200 1105 "-" "User-Agent: Firefly III Health Checker/2.3"
[2024-11-23 15:20:44] local.DEBUG: Now at App\Http\Controllers\TokenController::callback  
[2024-11-23 15:20:44] local.DEBUG: State is valid!  
[2024-11-23 15:20:44] local.DEBUG: Params for access token {"form_params":{"grant_type":"authorization_code","client_id":17,"redirect_uri":"http://firefly-import.raspi5.home.arpa/callback","code_verifier":"EqafPvGLJuFRpEFJy2jBrl3jgHoeVXRMdDcHlvufcB44ymq0QfenPmDAFg4kg1xrGhnUCTcIt0Q06WCXiRXbdEeYnhOX0ncI2woAksMxsrrH2waYstCKwD1hjlnh6WIK","code":"def50200c6711e2d83a0a467c8492e1fadf033da5bc07d07f08dfef9faf574c67d3b2e9eef7d000ce67777c7eacf96434c053101e7c4b1e3df6cffe9320648362e27cd5d82c8618f9b1cb97834483562521759e07e06a097c31140dd8060a567a4815c24dcab8a2759121e1212d889fc17a5a33a28273dc8630ee87650ee23af1057994aa7041179fb20cad85035c0b5d9abb237fd58731d0f845bc8ac37d8fe30ffa1062517b6885036b1bc7b27123fb60bea037e62dd17b6a2ccaf98cafafdd20fabc8ae7d037f2611e59c45fe74518d5074112ffd5e1858152753eb323521b25f77d64f9f5d2cf38932c886c884891e650bb63943fb45f82950d9e397f096a7210f758f9d2865c5a89bac5f3197b4edf845f47ca19c52727a908e55ca4bbb8ca5e709e0b690c281fae2bbb63ff1a23569983d574eb9e28584c757bafb08257f22a25f17b730b49870380ca6ed99bcd72dd671f13fe28906fb7e58c13df26833a4c99332d05b69c8def4103f26600cffe07765a8b288539c0dd138e84b248c9cad82adc05e4afd622b96929d80b19bcd16f536f447785d0ff92c"}} 
[2024-11-23 15:20:44] local.DEBUG: Will contact "http://app:8080/oauth/token" for a token.  
[2024-11-23 15:20:44] local.ERROR: Client exception when decoding response: Client error: `POST http://app:8080/oauth/token` resulted in a `401 Unauthorized` response:
{"error":"invalid_client","error_description":"Client authentication failed","message":"Client authentication failed"}
  
[2024-11-23 15:20:44] local.ERROR: Response from server: "{"error":"invalid_client","error_description":"Client authentication failed","message":"Client authentication failed"}"  
172.19.0.15 - - [23/Nov/2024:15:20:44 +0100] "GET /callback?code=def50200c6711e2d83a0a467c8492e1fadf033da5bc07d07f08dfef9faf574c67d3b2e9eef7d000ce67777c7eacf96434c053101e7c4b1e3df6cffe9320648362e27cd5d82c8618f9b1cb97834483562521759e07e06a097c31140dd8060a567a4815c24dcab8a2759121e1212d889fc17a5a33a28273dc8630ee87650ee23af1057994aa7041179fb20cad85035c0b5d9abb237fd58731d0f845bc8ac37d8fe30ffa1062517b6885036b1bc7b27123fb60bea037e62dd17b6a2ccaf98cafafdd20fabc8ae7d037f2611e59c45fe74518d5074112ffd5e1858152753eb323521b25f77d64f9f5d2cf38932c886c884891e650bb63943fb45f82950d9e397f096a7210f758f9d2865c5a89bac5f3197b4edf845f47ca19c52727a908e55ca4bbb8ca5e709e0b690c281fae2bbb63ff1a23569983d574eb9e28584c757bafb08257f22a25f17b730b49870380ca6ed99bcd72dd671f13fe28906fb7e58c13df26833a4c99332d05b69c8def4103f26600cffe07765a8b288539c0dd138e84b248c9cad82adc05e4afd622b96929d80b19bcd16f536f447785d0ff92c&state=rs6o7IScd722pNQoJFd6Lu4szVEpcICbIyipyMm1 HTTP/1.1" 200 1165 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
```


I don't really get why I get a client error. From my knowledge and I what I understand of Docker, I should be able to reach between containers using the http://app:8080, but something seems off here. Pinging the app from the importer seems to work:
```
docker exec -it firefly_iii_importer curl -I http://app:8080
HTTP/1.1 302 Found
Date: Sat, 23 Nov 2024 14:39:56 GMT
Server: Apache
Cache-Control: no-cache, private
Location: http://app:8080/login
X-Frame-Options: deny
Content-Security-Policy: default-src 'none'; object-src 'none'; script-src 'unsafe-eval' 'strict-dynamic' 'nonce-fNLXnyMPMu3lvxhXlCFtCg=='; style-src 'unsafe-inline' 'self'; base-uri 'self'; font-src 'self' data:; connect-src 'self' ; img-src 'self' data: 'nonce-fNLXnyMPMu3lvxhXlCFtCg==' ; manifest-src 'self'; form-action 'self'
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Referrer-Policy: no-referrer
X-Permitted-Cross-Domain-Policies: none
X-Robots-Tag: none
Feature-Policy: geolocation 'none'; midi 'none'; sync-xhr 'self'; microphone 'none'; camera 'none'; magnetometer 'none'; gyroscope 'none'; fullscreen 'self'; payment 'none'
Set-Cookie: XSRF-TOKEN=eyJpdiI6IkU1ampURU8xaHZsQmM3TEFRZmlBSkE9PSIsInZhbHVlIjoiVHVzTytyV1dkSm9qQjZzTmNCQWJnMlF6R1RMZG1VeWVjci9iaXdQYkszQTY3QkxoRGZqSm56V3RJR29EY29ESzlLYjZRZjU2cXBhdTBocVl0UXhxOW1XcjlWVDJCNEpHSG5nM0EyN0djS2ZNTlI3SktmNGpHWW9GZnJheFVRL2giLCJtYWMiOiIwYTgwOTBjOGM4OTc0MzI0YjQxMjM3MmM2NDIyZWJmNWUyZjMxNTM4MGVjNTIzZGQyMzM3MjFlNDVkZjAxYzc2IiwidGFnIjoiIn0%3D; expires=Sat, 23 Nov 2024 16:39:56 GMT; Max-Age=7200; path=/; samesite=lax
Set-Cookie: firefly_session=eyJpdiI6IkVEcld2enQ1emQ5dXdPdGFHOVEwVlE9PSIsInZhbHVlIjoiRjlNRkU2aWFrWmgzS3d2MEFxNFdLczhoeDBXV21TNFZXNTVkNVNob3E5dHJaQys1ZHN2citBMjAxT1BBbzQ3dWRYckxJaThwQ2swNllnQmRDYTdqQlZzb0NVM1hRQ3ZSMzdDRVo3VC9wTk5Ya2I3RzNiTXhnTnR3bnUxSXpIM1IiLCJtYWMiOiIxODNjYjU1ZDc0Njk5OTI2ZTNjYjQyZTVlMWY0OGIzMGU2ZDZlYjAxYTZjN2E5YjQ4ZmI3MWI2MGNiYmJlZDM1IiwidGFnIjoiIn0%3D; path=/; httponly; samesite=lax
Content-Type: text/html; charset=utf-8


```

 I've tried with :
- the public url of the Firefly III instance too, no luck.
- localhost:8080, no luck
- localhost without ports, no luck

I'm pretty sure it's a configuration issue with reverse proxy and not the app itself, as my installation was working before (but not properly a reverse proxy, nginx proxy manager with the main app proxied, but not the importer) 

I'm pretty sure the solution is simple and will look more, will keep updated if I find a solution. If someone have a clue...

Best,

**Reply by Armgd** [reply]
Ran a different scenario, if I use a Personal Access token instead, setting:
```
FIREFLY_III_CLIENT_ID=
FIREFLY_III_ACCESS_TOKEN=ey...
```

it works without any issue. So I suspect either a reverse proxy issue not passing parameters correctly when using the Oauth way.

**Comment by codehacker84** [comment]
same issue

**Comment by JoeSHMOEK** [comment]
I had this issue earlier on my K8s deployment and it appears to be due to TLS Verification.

**Comment by JC5** [comment]
The new release, which you can already try when you use :develop, should no longer display "empty".
